src/wasm/wasm-ir-builder.cpp

```diff
@@ -0,0 +1,921 @@
+// }
+
+// Result<> IRBuilder::makeCallIndirect() {
+//   auto* curr = wasm.allocator.alloc<CallIndirect>();
+//   return visitCallIndirect(curr);
+// }
+
+Result<> IRBuilder::makeLocalGet(Index local) {
+  if (!func) {
+    return Err{"local.get must be inside a function"};
+  }
+  if (local >= func->getNumLocals()) {
+    return Err{"local index out of bounds"};
+  }
+  auto* curr = wasm.allocator.alloc<LocalGet>();
+  curr->type = func->getLocalType(local);
+  curr->index = local;
+  return visitLocalGet(curr);
+}
+
+Result<> IRBuilder::makeLocalSet(Index local) {
+  if (!func) {
+    return Err{"local.set must be inside a function"};
+  }
+  if (local >= func->getNumLocals()) {
+    return Err{"local index out of bounds"};
+  }
+  auto* curr = wasm.allocator.alloc<LocalSet>();
+  curr->index = local;
+  return visitLocalSet(curr);
+}
+
+Result<> IRBuilder::makeLocalTee(Index local) {
+  if (!func) {
+    return Err{"local.tee must be inside a function"};
+  }
+  if (local >= func->getNumLocals()) {
+    return Err{"local index out of bounds"};
+  }
+  auto* curr = wasm.allocator.alloc<LocalSet>();
+  curr->type = func->getLocalType(local);
+  curr->index = local;
+  return visitLocalSet(curr);
+}
+
+Result<> IRBuilder::makeGlobalGet(Name global) {
+  if (!wasm.getGlobalOrNull(global)) {
+    return Err{"global '"s + global.toString() + "' does not exist"};
+  }
+  auto* curr = wasm.allocator.alloc<GlobalGet>();
+  curr->type = wasm.getGlobal(global)->type;
+  curr->name = global;
+  return visitGlobalGet(curr);
+}
+
+Result<> IRBuilder::makeGlobalSet(Name global) {
+  if (!wasm.getGlobalOrNull(global)) {
+    return Err{"global '"s + global.toString() + "' does not exist"};
+  }
+  auto* curr = wasm.allocator.alloc<GlobalSet>();
+  curr->name = global;
+  return visitGlobalSet(curr);
+}
+
+Result<> IRBuilder::makeLoad(unsigned bytes,
+                             bool signed_,
+                             Address offset,
+                             unsigned align,
+                             Type type,
+                             Name mem) {
+  auto* curr = wasm.allocator.alloc<Load>();
+  curr->isAtomic = false;
+  curr->bytes = bytes;
+  curr->signed_ = signed_;
+  curr->offset = offset;
+  curr->align = align;
+  curr->type = type;
+  curr->memory = mem;
+  return visitLoad(curr);
+}
+
+Result<> IRBuilder::makeStore(
+  unsigned bytes, Address offset, unsigned align, Type type, Name mem) {
+  auto* curr = wasm.allocator.alloc<Store>();
+  curr->isAtomic = false;
+  curr->bytes = bytes;
+  curr->offset = offset;
+  curr->align = align;
+  curr->valueType = type;
+  curr->memory = mem;
+  return visitStore(curr);
+}
+
+Result<>
+IRBuilder::makeAtomicLoad(unsigned bytes, Address offset, Type type, Name mem) {
+  auto* curr = wasm.allocator.alloc<Load>();
+  curr->isAtomic = true;
+  curr->bytes = bytes;
+  curr->signed_ = false;
+  curr->offset = offset;
+  curr->align = bytes;
+  curr->type = type;
+  curr->memory = mem;
```

Comment:
Reading long sequences like these makes me wonder if we can share code with Builder somehow. Something like this:
```cpp
Result<>
IRBuilder::makeAtomicLoad(unsigned bytes, Address offset, Type type, Name mem) {
  return visitLoad(builder.makeLoad(nullptr, bytes, offset, type, mem));
}
```
That is, call the Builder method and pass it nullptr for the `Expression*` children. Of course this won't work though as it calls `finalize` there.

The reverse might work, though?
```cpp
Result<>
IRBuilder::makeAtomicLoad(unsigned bytes, Address offset, Type type, Name mem) {
  auto* child = getChild();
  builder.makeLoad(child, bytes, offset, type, mem));
}
```
That is, get the children first using some new code (`getChild()` gets one child, another method could get an arbitrary number, etc.), and then pass that into Builder. It seems like that would also avoid the need for the delegates stuff in `visitExpression`?

A downside to that is we need to write out the number of children here explicitly. But it still seems like significantly less code overall. Does this make sense or am I missing something?

Reply:
Yeah, it would be nice to share this code. Unfortunately I couldn't figure out a good way to do so.

As you mentioned, using `builder` to do the initial allocation and field setting wouldn't work because the children would not yet be initialized.

Popping the children up front and then calling into `builder` would work, and is what wat-parser.cpp did before this change, but has the trade off that it doesn't benefit from being able to simply call `visit` (or `visitXYZ`) to install the children. Popping all the children manually is also a nontrivial amount of code that I was happy to be able to remove, and we benefit from having `visit` be well-tested because that's the API that @ashleynh will be using for outlining.

Somewhat interestingly, it would be more straightforward to reimplement `Builder` in terms of `IRBuilder`, but of course that would do a lot of unnecessary work compared to the current implementation.

Reply:
One could imagine having a separate, shared utility called `ImmediateBuilder` that allocates nodes and sets their non-child fields and does nothing else. That could be shared between both `Builder` and `IRBuilder`, but is otherwise completely useless, so I don't think it would be worth it.

Reply:
> Popping the children up front and then calling into builder would work, and is what wat-parser.cpp did before this change, but has the trade off that it doesn't benefit from being able to simply call visit (or visitXYZ) to install the children. Popping all the children manually is also a nontrivial amount of code that I was happy to be able to remove

Can we automate that by using delegations in a helper function that gets an Expression type and pops the right number of children, returning them in a vector? The only special cases would need to be optional children, but those already need to be special cased in this code.

Reply:
PTAL at the latest version that `visit`s stack-allocated `Expression`s to get the children, then calls `builder.makeXYZ` to arena-allocate, initialize, and finalize the returned node.